lib/std/mem.zig

```diff
@@ -258,6 +258,38 @@ pub fn zeroes(comptime T: type) T {
     }
 }
 
+/// Compares two slices in constant time (for a given length) and returns whether they are equal.
+/// This function was designed to compare short cryptographic secrets (MACs, signatures).
+/// For all other applications, use mem.eql() instead.
+pub fn timingSafeEql(comptime T: type, a: T, b: T) bool {
+    switch (@typeInfo(T)) {
+        .Array => {
+            const C = @typeInfo(T).Array.child;
+            var acc = @as(C, 0);
+            for (a) |x, i| {
+                acc |= x ^ b[i];
+            }
+            comptime const s = @typeInfo(C).Int.bits;
+            comptime const Cu = std.meta.Int(.unsigned, s);
+            comptime const Cext = std.meta.Int(.unsigned, s + 1);
+            return @bitCast(bool, @truncate(u1, (@as(Cext, @bitCast(Cu, acc)) -% 1) >> s));
```

Comment:
Shouldn't `acc >> i` rather be `acc >> 1`?

The code this generates is awful :( And I really don't think it is necessary.

Reply:
> Shouldn't acc >> i rather be acc >> 1?

It's repeatedly folding `acc` in half, `i` is the midpoint.

> The code this generates is awful :(

That's the spirit, keep the optimizer away with bizarre constructions.

> And I really don't think it is necessary.

Maybe that's too much, yes :)

Reply:
Fair enough.
This could be slightly simpler as
```zig
return @subWithOverflow(Cu, @bitCast(Cu, acc, 1, &unused);
```
if you want to use the carry bit as Boolean flag.

Reply:
Using the carry bit may increase the chances of having this turned into a conditional jump :)

Reply:
...the code you wrote is doing the very same thing?

Reply:
Not really. What is returned from the function can be directly the result of a shift+and operation. Converting a flag to a general purpose register may require a conditional branch.

Reply:
> Converting a flag to a general purpose register may require a conditional branch.

An add/subtract with carry is enough...

Reply:
The code generated by `@subWithOverflow` is always worse.

- x86_64: [usual method](https://godbolt.org/z/Y4o8xe) - [subWithOverflow](https://godbolt.org/z/rGnnon)
- aarch64: [usual method](https://godbolt.org/z/8xx6fb) - [subWithOverflow](https://godbolt.org/z/h8rvj5)
- mips: [usual method](https://godbolt.org/z/6KqrrG) - [subWithOverflow](https://godbolt.org/z/cMsdhe)

Reply:
> The code generated by @subWithOverflow is always worse.

And nobody said otherwise?
The suggestion was to make the code slightly cleaner and remove the double cast, but I'll take your word on the code being _worse_ and take it back.
LLVM should learn how to optimize the overflowing ops when the result is discarded, something new for my backlog.

Just for fun, try comparing the generated code for `[16]usize` arrays on x86_64/x86 (that's representative of a comparison between two math.big.int limbs) :)